File: src/easy/Q0100.py
"""
100. Same Tree
https://leetcode.com/problems/same-tree/
"""
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# O(n) time, O(log n) space, where n is the maximum number of nodes in p or q.
# Assumes p and q are ~balanced trees. If p or q are fully imbalanced then the space
# complexity is O(n).
def is_same_tree(p: Optional[TreeNode], q: Optional[TreeNode]):
    if not (p or q):
        return True
    if not (p and q):
        return False
    if p.val != q.val:
        return False
    return is_same_tree(p.left, q.left) and is_same_tree(p.right, q.right)

File: src/easy/Q0100/Q0100.cpp
// 100. Same Tree
// https://leetcode.com/problems/same-tree/

#include <functional>

using namespace std;

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
};

// Time: O(n), Space: O(log n), where n is the number of nodes in `p` or `q`,
// whichever has more. This assumes p and q and ~balanced binary trees.
// Otherwise, the space complexity would be O(n).
auto is_same_tree(TreeNode* p, TreeNode*) -> bool {
  function<bool(TreeNode*, TreeNode*)> dfs = [&dfs](TreeNode* p, TreeNode* q) {
    if (!p and !q) return true;
    if (!p or !q) return false;
    if (p->val != q->val) return false;
    return dfs(p->left, q->left) and dfs(p->right, q->right);
  };
  return dfs(p, q);
}
